network_simulator/statistics.py

```diff
@@ -4,6 +4,200 @@
+    :ivar list packet_rec_times: a list of (timestamp, packet_size)
+    tuples. packet_size is in bits. Time is in seconds.
+    """
+    def __init__(self, packet_sent_times=[], packet_rec_times=[]):
+        packet_sent_times = packet_sent_times
+        packet_rec_times = packet_rec_times
+
+
+
 class Statistics(object):
-    # TODO(sharon): implement.
-    pass
\ No newline at end of file
+    """
+    Intended to be owned by the main loop to record all statistics within
+    the network.
+
+    :ivar dict link_stats: a map from (end_1_id, end_2_id) to
+    :class:`.LinkStats`. end_id's are strings.
+    :ivar dict flow_stats: a map from flow ID to :class:`.FlowStats`. flow
+    ids are strings.
+    :ivar dict host_stats: a map from host ID to :class:`.HostStats`. host
+    ids are strings.
+    """
+    def __init__(self, link_stats={}, flow_stats={}, host_stats={}):
+        self.link_stats = link_stats
+        self.flow_stats = flow_stats
+        self.host_stats = host_stats
+
+    def get_link_stats(self, link):
+        """
+        Check whether a link has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Link link: link to be determined.
+        :return :class:`.LinkStats`.
+        """
+        # Create a tuple of the pair and check if it exists in the dict.
+        tup = (link.end_1_addr, link.end_2_addr)
+
+        if tup not in self.link_stats:
+            self.link_stats[tup] = LinkStats()
+        return self.link_stats[tup]
+
+    def get_flow_stats(self, flow):
+        """
+        Check whether a flow has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Flow flow: flow to be determined.
+        :return :class:`.FlowStats`.
+        """
+        if flow.flow_id not in self.flow_stats:
+            self.flow_stats[flow.flow_id] = FlowStats()
+        return self.flow_stats[flow.flow_id]
+
+    def get_host_stats(self, host):
+        """
+        Check whether a host has already existed in the dictionary; if so,
+        skip the rest of the steps and return the stats object. If not,
+        create a new mapping then return the stats object.
+
+        :param Host host: host to be determined.
+        :return :class:`.HostStats`.
+        """
+        if host.address not in self.host_stats:
+            self.host_stats[host.address] = HostStats()
+        return self.host_stats[host.address]
+
+    def link_buffer_occ_change(self, link, curr_time):
+        """
+        Update the buffer size as a link gains or loses a queue element.
+
+        :param Link link: link that has a change of buffer size.
+        """
+        stats = self.get_link_stats(link)
+        # TODO(sharon): Check with Cody if qsize() would work in this case.
```

Comment:
What we need to check is whether we want to count RoutingPackets as well, not if qsize() will work.